src/net.cpp

```diff
@@ -362,6 +362,20 @@ bool IsLocal(const CService& addr)
     return mapLocalHost.count(addr) > 0;
 }
 
+static void InitializePermissionFlags(NetPermissionFlags& flags, ServiceFlags& service_flags) {
+    if (NetPermissions::HasFlag(flags, NetPermissionFlags::Implicit)) {
+        NetPermissions::ClearFlag(flags, NetPermissionFlags::Implicit);
+        if (gArgs.GetBoolArg("-whitelistforcerelay", DEFAULT_WHITELISTFORCERELAY)) NetPermissions::AddFlag(flags, NetPermissionFlags::ForceRelay);
+        if (gArgs.GetBoolArg("-whitelistrelay", DEFAULT_WHITELISTRELAY)) NetPermissions::AddFlag(flags, NetPermissionFlags::Relay);
+        NetPermissions::AddFlag(flags, NetPermissionFlags::Mempool);
+        NetPermissions::AddFlag(flags, NetPermissionFlags::NoBan);
+    }
+
+    if (NetPermissions::HasFlag(flags, NetPermissionFlags::BloomFilter)) {
+        service_flags = static_cast<ServiceFlags>(service_flags | NODE_BLOOM);
+    }
```

Comment:
I might be misunderstanding something here, but I think this chunk can be moved to `net_processing.cpp` `InitializeNode()` because:
- the `CNode` will already have its `m_permission_flags` set in `ConnectNode()`
- `m_permission_flags` is all that's needed to set `our_services` inside `InitializeNode()`
- then you won't need to use `std::optional<std::pair<CNode*, ServiceFlags>> ` which is, hey, a real banger of a type!

Otherwise I don't see a reason to compute and then pass around those service flags through all those functions before they are used.  (`ConnectNode()->OpenNetworkConnection()->InitializeNode()`)

As a side note, this looks like a lot of work to allow BLOOM for an *outgoing* peer, which must be a really weird edge case anyway right?

Reply:
> `m_permission_flags` is all that's needed to set `our_services` inside `InitializeNode()`

Are you sure? I think we need to access of `nLocalServices` for it and we need to do it in `CConnman`, doesn't it?

Reply:
Hi sorry just getting back to this review. Check out my top two commits here: 

https://github.com/pinheadmz/bitcoin/commits/zip-27114

I'm still running the tests locally and I could be totally missing something, but this was what I was thinking. I don't think `ConnectNode()` needs to return a pair with the service flags, as long as we keep your other changes that pass the *permission flags* along inside the CNode

Reply:
done!